DQM/BeamMonitor/plugins/BeamSpotDipServer.cc

```diff
@@ -0,0 +1,740 @@
+
+  subjectCMS = ps.getUntrackedParameter<string>("subjectCMS"); 
+  subjectLHC = ps.getUntrackedParameter<string>("subjectLHC"); 
+  subjectPV  = ps.getUntrackedParameter<string>("subjectPV" ); 
+
+  readFromNFS = ps.getUntrackedParameter<bool>("readFromNFS");
+
+  sourceFile  = ps.getUntrackedParameter<string>("sourceFile" ); // beamspot
+  sourceFile1 = ps.getUntrackedParameter<string>("sourceFile1"); // tk status
+
+  timeoutLS = ps.getUntrackedParameter<vector<int>>("timeoutLS");
+
+  //
+  bsLegacyToken_ = esConsumes<edm::Transition::EndLuminosityBlock>();
+
+  dcsStatus_ = consumes<DcsStatusCollection>(
+    ps.getUntrackedParameter<string>("DCSStatus", "scalersRawToDigi")); 
+
+  //
+  dip = Dip::create("CmsBeamSpotServer");
+  cerr << "[Dip] reading from " << (readFromNFS ? "file (NFS)" : "database")
+       << endl;
+}
+
+/*****************************************************************************/ 
+void BeamSpotDipServer::bookHistograms(
+  DQMStore::IBooker&, const edm::Run&, const edm::EventSetup&)
+{
+  // do nothing
+}
+
+/*****************************************************************************/ 
+void BeamSpotDipServer::dqmBeginRun(
+  const edm::Run& r, const edm::EventSetup&)
+{
+  cout << "[Dip] begin run " << r.run() << endl;
+
+  try
+  {
+    ErrHandler errHandler;
+
+    cout << "[Dip] server started at " + getDateTime() << endl;
+
+    cout << "[Dip] creating publication " + subjectCMS << endl;
+    publicationCMS = dip->createDipPublication(subjectCMS.c_str(),&errHandler);
+        messageCMS = dip->createDipData();
+
+    cout << "[Dip] creating publication " + subjectLHC << endl;
+    publicationLHC = dip->createDipPublication(subjectLHC.c_str(),&errHandler);
+        messageLHC = dip->createDipData();
+
+    cout << "[Dip] creating publication " + subjectPV << endl;
+    publicationPV  = dip->createDipPublication(subjectPV.c_str(), &errHandler);
+        messagePV  = dip->createDipData();
+
+    trueRcd(); // starts with all 0
+    publishRcd("UNINITIALIZED","",true,false);
+  }
+  catch (exception & e)
+  {
+    cerr << "[Dip] exception (start up): " << e.what() << endl;
+  }
+
+  quality = qualities[0]; // start with Uncertain
+}
+
+/*****************************************************************************/ 
+void BeamSpotDipServer::dqmBeginLuminosityBlock(
+  const edm::LuminosityBlock&, const edm::EventSetup&)
+{
+  // do nothing
+}
+
+/*****************************************************************************/ 
+void BeamSpotDipServer::analyze(
+  const edm::Event& iEvent, const edm::EventSetup&)
+{
+  if(!readFromNFS)
+  {
+    // get runnumber
+    runnum = iEvent.run();
+
+    // get tracker status if in a new lumisection
+    int nthlumi = iEvent.luminosityBlock();
+
+    if (nthlumi > lastlumi) { // check every LS
+      lastlumi = nthlumi;
+
+      edm::Handle<DcsStatusCollection> dcsStatus;
+      iEvent.getByToken(dcsStatus_, dcsStatus);
```

Comment:
isn't this going to go unavailable in Run3?
https://indico.cern.ch/event/791284/contributions/3305157/attachments/1810322/2956239/CMS_Online_Monitoring_System_Lyon1.pdf#page=27

Reply:
I see. Then I would need to find out where TkStatus (status of the tracker subsystems) would be avaialble in the future.

Reply:
Based on the [document linked from the slides](https://docs.google.com/document/d/1qV72Vj0IMGN7cX33B12xKzLII6NEM2gSUi4fn995bRQ/edit#heading=h.9ifa66mb6oxt) this:
```C++
 #include "DataFormats/OnlineMetaData/interface/DCSRecord.h"
 ...
  dcsRecordInputTag_ = config.getParameter<edm::InputTag>("dcsRecordInputTag");
  dcsRecordToken_ = consumes<DCSRecord>(dcsRecordInputTag_);
...
  edm::Handle<DCSRecord> dcsRecord;
  event.getByToken(dcsRecordToken_, dcsRecord);
```
and feeding:
```python
     dcsRecordInputTag = cms.InputTag ( "onlineMetaDataDigis" ),
```

should work.

Reply:
Thanks for the information, I have replaced DCSStatus with DCSRecord.